client/starwhale/api/_impl/data_store.py

```diff
@@ -158,21 +369,33 @@ def __str__(self) -> str:
 
 _TYPE_NAME_DICT = {str(v): v for v in _TYPE_DICT.values()}
 
-_ID_CUSTOM_TYPE_DICT: Dict[str, Type[Any]] = {}
-
-_CUSTOM_TYPE_ID_DICT: Dict[Type[Any], str] = {}
 
+def _get_type(obj: Any) -> SwType:
+    if isinstance(obj, list):
+        element_type: SwType = UNKNOWN
+        for element in obj:
+            element_type = element_type.merge(_get_type(element))
+        return SwListType(element_type)
+    if isinstance(obj, SwObject):
+        attrs = {}
+        for k, v in obj.__dict__.items():
+            attrs[k] = _get_type(v)
+        return SwObjectType(type(obj), attrs)
+    ret = _TYPE_DICT.get(type(obj), None)
+    if ret is None:
+        raise RuntimeError(f"unsupported type {type(obj)}")
+    return ret
 
-def _get_type(obj: Any) -> Optional[SwType]:
-    return _TYPE_DICT.get(type(obj), None)
 
+class SwObject:
+    def __str__(self) -> str:
+        return json.dumps(self.__dict__)
 
-def register_custom_type(id: str, type: Type[Any]) -> None:
-    _ID_CUSTOM_TYPE_DICT[id] = type
-    _CUSTOM_TYPE_ID_DICT[type] = id
+    def __eq__(self, other: Any) -> bool:
+        return type(other) is type(self) and self.__dict__ == other.__dict__
```

Comment:
do we need a deep equal check for the class __dict__?

Reply:
I guess == is enough for current supported types